lighthouse-core/gather/gatherers/image-elements-snapshot.js

```diff
@@ -0,0 +1,376 @@
+      element.cssEffectiveRules = {width, height, aspectRatio};
+    } catch (err) {
+      if (/No node.*found/.test(err.message)) return;
+      throw err;
+    }
+  }
+
+  /**
+   *
+   * @param {LH.Gatherer.FRTransitionalDriver} driver
+   * @param {LH.Artifacts.ImageElement[]} elements
+   * @param {Record<string, LH.Artifacts.NetworkRequest>=} indexedNetworkRecords
+   */
+  async collectExtraDetails(driver, elements, indexedNetworkRecords) {
+    // Don't do more than 5s of this expensive devtools protocol work. See #11289
+    let reachedGatheringBudget = false;
+    setTimeout(_ => (reachedGatheringBudget = true), 5000);
+    let skippedCount = 0;
+
+    for (const element of elements) {
+      // Pull some of our information directly off the network record.
+      const networkRecord = indexedNetworkRecords && indexedNetworkRecords[element.src];
+      element.mimeType = networkRecord && networkRecord.mimeType;
+
+      if (reachedGatheringBudget) {
+        skippedCount++;
+        continue;
+      }
+
+      // Need source rules to determine if sized via CSS (for unsized-images).
+      if (!element.isInShadowDOM && !element.isCss) {
+        await this.fetchSourceRules(driver.defaultSession, element.node.devtoolsNodePath, element);
+      }
+      // Images within `picture` behave strangely and natural size information isn't accurate,
+      // CSS images have no natural size information at all. Try to get the actual size if we can.
+      // Additional fetch is expensive; don't bother if we don't have a networkRecord for the image.
+      if (element.isPicture || element.isCss || element.srcset) {
+        if (indexedNetworkRecords && !networkRecord) continue;
+        await this.fetchElementWithSizeInformation(driver, element);
+      }
+    }
+
+    if (reachedGatheringBudget) {
+      log.warn('ImageElements', `Reached gathering budget of 5s. Skipped extra details for ${skippedCount}/${elements.length}`); // eslint-disable-line max-len
+    }
+  }
+
+  /**
+   * @param {LH.Gatherer.FRTransitionalContext} context
+   * @param {Record<string, LH.Artifacts.NetworkRequest>=} indexedNetworkRecords
+   * @return {Promise<LH.Artifacts['ImageElements']>}
+   */
+  async _getArtifact(context, indexedNetworkRecords) {
+    const session = context.driver.defaultSession;
+    const executionContext = context.driver.executionContext;
+
+    const elements = await executionContext.evaluate(collectImageElementInfo, {
+      args: [],
+      deps: [
+        pageFunctions.getElementsInDocumentString,
+        pageFunctions.getBoundingClientRectString,
+        pageFunctions.getNodeDetailsString,
+        getClientRect,
+        getPosition,
+        getHTMLImages,
+        getCSSImages,
+      ],
+    });
+
+    await Promise.all([
+      session.sendCommand('DOM.enable'),
+      session.sendCommand('CSS.enable'),
+      session.sendCommand('DOM.getDocument', {depth: -1, pierce: true}),
+    ]);
+
```

Comment:
Element sorting was removed for convenience of making this draft. Will need to restore it somehow before landing.